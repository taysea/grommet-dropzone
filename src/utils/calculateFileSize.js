const SIZE = {
  B: "bytes",
  KB: "KB",
  MB: "MB",
  GB: "GB"
};

export const calculateFileSize = size => {
  let adjustedSize = size;
  let count = 0;
  let suffix;

  while (adjustedSize > 1000) {
    adjustedSize = adjustedSize / 1000;
    count = count + 1;
  }
  if (count === 0) {
    suffix = SIZE.B;
  } else if (count === 1) {
    suffix = SIZE.KB;
  } else if (count === 2) {
    suffix = SIZE.MB;
  } else {
    suffix = SIZE.GB;
  }

  const calculatedSize = {
    value: adjustedSize,
    suffix: suffix
  };
  return calculatedSize;
};
