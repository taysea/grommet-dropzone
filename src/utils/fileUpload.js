export function fileUpload({ firstName, lastName, files }) {
  let formData = new FormData();

  files.forEach(name => {
    formData.append(name, files[name]);
  });
  formData.append(firstName, firstName);
  formData.append(lastName, lastName);

  // adjust the url below with your own
  return fetch("https://my-website.com/api/file/upload", {
    method: "POST",
    // if your app is storing auth tokens in a cookie include credentials
    // credentials: 'include',
    body: formData
  })
    .then(response =>
      response.json().then(json => ({
        status: response.status,
        statusText: response.statusText,
        json
      }))
    )
    .then(({ status, statusText, json }) => {
      if (status >= 400) {
        // API returned a crappy response
        console.log("error:", status, statusText, json);
      } else {
        // Upload done!
        return json;
      }
    })
    .catch(err => console.log(err));
}
