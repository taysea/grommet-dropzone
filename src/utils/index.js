export * from "./calculateFileSize";
export * from "./fileUpload";
export * from "./truncateText";
