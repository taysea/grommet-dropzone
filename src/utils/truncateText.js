export function truncateText(text) {
  if (text.length > 15) {
    const truncatedText = text.substr(0, 15);
    return truncatedText + "...";
  }
  return text;
}
