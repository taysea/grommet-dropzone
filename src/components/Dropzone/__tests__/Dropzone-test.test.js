import React from "react";
import {
  cleanup,
  fireEvent,
  render,
  waitForElement
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Dropzone } from "../";
import { calculateFileSize, truncateText } from "../../../utils";

describe("Dropzone", () => {
  let smallImage;
  let largeImage;
  window.URL.createObjectURL = jest.fn();

  beforeEach(() => {
    smallImage = createFile("gremlin.png", 100, "image/png");
    largeImage = createFile("gremlin.png", 10000, "image/png");
  });

  afterEach(() => {
    cleanup();
    window.URL.createObjectURL.mockReset();
  });

  test("renders basic dropzone", () => {
    const { container } = render(<Dropzone />);
    expect(container).toMatchSnapshot();
  });

  test("renders correct starting text for single file", () => {
    const { container, getByText } = render(<Dropzone />);
    const dropText = getByText("a file", { exact: false });
    expect(dropText).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("renders correct starting text for multiple files", () => {
    const { container, getByText } = render(<Dropzone multiple />);
    const dropText = getByText("files", { exact: false });
    expect(dropText).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("accept single file by default", async () => {
    const { container, getByText } = render(<Dropzone />);

    const dropzoneInput = getByText("Drag and drop", {
      exact: false
    }).previousSibling;

    setUpFileDrop(dropzoneInput, smallImage);

    fireEvent.drop(dropzoneInput);
    const acceptedFile = await waitForElement(() => getByText(smallImage.name));

    expect(acceptedFile).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("renders file size and calculates correct suffix", async () => {
    const { container, getByText } = render(<Dropzone showFileSize />);

    const dropzoneInput = getByText("Drag and drop", {
      exact: false
    }).previousSibling;

    setUpFileDrop(dropzoneInput, largeImage);

    fireEvent.drop(dropzoneInput);

    const acceptedFile = await waitForElement(() => getByText(smallImage.name));
    expect(acceptedFile).toBeInTheDocument();

    const fileSize = calculateFileSize(largeImage.size);
    expect(
      getByText(`${fileSize.value}.0 ${fileSize.suffix}`)
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("renders file preview", async () => {
    const { container, getByText } = render(<Dropzone showPreview />);

    const dropzoneInput = getByText("Drag and drop", {
      exact: false
    }).previousSibling;
    setUpFileDrop(dropzoneInput, smallImage);

    fireEvent.drop(dropzoneInput);
    const acceptedFile = await waitForElement(() => getByText(smallImage.name));

    expect(acceptedFile).toBeInTheDocument();
    expect(container.querySelector("img")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("renders remove button", async () => {
    const { container, getByLabelText, getByText } = render(<Dropzone />);

    const dropzoneInput = getByText("Drag and drop", {
      exact: false
    }).previousSibling;

    setUpFileDrop(dropzoneInput, smallImage);

    fireEvent.drop(dropzoneInput);
    const acceptedFile = await waitForElement(() => getByText(smallImage.name));

    expect(acceptedFile).toBeInTheDocument();
    expect(getByLabelText("FormClose")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("truncates text for long file name", async () => {
    const { container, getByText } = render(<Dropzone />);

    const dropzoneInput = getByText("Drag and drop", {
      exact: false
    }).previousSibling;

    const longTitleFile = createFile(
      "very-long-image-title-that-should-be-truncated.png",
      100,
      "image/png"
    );

    const truncatedTitle = truncateText(longTitleFile.name);
    setUpFileDrop(dropzoneInput, longTitleFile);

    fireEvent.drop(dropzoneInput);
    const acceptedFile = await waitForElement(() => getByText(truncatedTitle));

    expect(acceptedFile).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("removes file when remove button is clicked", async () => {
    const { container, getByLabelText, getByText } = render(<Dropzone />);

    const dropzoneInput = getByText("Drag and drop", {
      exact: false
    }).previousSibling;

    setUpFileDrop(dropzoneInput, smallImage);

    fireEvent.drop(dropzoneInput);
    const acceptedFile = await waitForElement(() => getByText(smallImage.name));
    expect(acceptedFile).toBeInTheDocument();

    fireEvent.click(getByLabelText(`remove ${smallImage.name}`));
    expect(getByText("a file", { exact: false })).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});

function setUpFileDrop(dropzoneInput, file) {
  Object.defineProperty(dropzoneInput, "files", {
    value: [file]
  });
}

// This function is borrowed from react-dropzone tests
function createFile(name, size, type) {
  const file = new File([], name, { type });
  Object.defineProperty(file, "size", {
    get() {
      return size;
    }
  });
  return file;
}
