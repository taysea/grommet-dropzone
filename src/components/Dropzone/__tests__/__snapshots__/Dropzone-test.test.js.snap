// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropzone accept single file by default 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 icmjSc"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="StyledBox-sc-13pk1d4-0 eEjmTS"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eQTSO"
      >
        <span
          class="StyledText-sc-1sadyjn-0 bnqisW"
        >
          gremlin.png
        </span>
      </div>
      <button
        aria-label="remove gremlin.png"
        class="StyledButton-sc-323bzc-0 kQcSJk"
        type="button"
      >
        <svg
          aria-label="FormClose"
          class="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7 L17,17 M7,17 L17,7"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Dropzone removes file when remove button is clicked 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 doFbKC"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <span
      class="StyledText-sc-1sadyjn-0 fWSbXS"
    >
      Drag and drop or 
      <a
        class="StyledAnchor-sc-1rp7lwl-0 jgeOSx"
      >
        choose
      </a>
       
      a file
    </span>
  </div>
</div>
`;

exports[`Dropzone renders basic dropzone 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 doFbKC"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <span
      class="StyledText-sc-1sadyjn-0 fWSbXS"
    >
      Drag and drop or 
      <a
        class="StyledAnchor-sc-1rp7lwl-0 jgeOSx"
      >
        choose
      </a>
       
      a file
    </span>
  </div>
</div>
`;

exports[`Dropzone renders correct starting text for multiple files 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 doFbKC"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <span
      class="StyledText-sc-1sadyjn-0 fWSbXS"
    >
      Drag and drop or 
      <a
        class="StyledAnchor-sc-1rp7lwl-0 jgeOSx"
      >
        choose
      </a>
       
      files
    </span>
  </div>
</div>
`;

exports[`Dropzone renders correct starting text for single file 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 doFbKC"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <span
      class="StyledText-sc-1sadyjn-0 fWSbXS"
    >
      Drag and drop or 
      <a
        class="StyledAnchor-sc-1rp7lwl-0 jgeOSx"
      >
        choose
      </a>
       
      a file
    </span>
  </div>
</div>
`;

exports[`Dropzone renders file preview 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 icmjSc"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="StyledBox-sc-13pk1d4-0 eEjmTS"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eQTSO"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 bCvkuC"
        >
          <img
            class="StyledImage-ey4zx9-0 UumnI"
          />
        </div>
        <div
          class="StyledBox__StyledBoxGap-sc-13pk1d4-1 gufrXq"
        />
        <span
          class="StyledText-sc-1sadyjn-0 bnqisW"
        >
          gremlin.png
        </span>
      </div>
      <button
        aria-label="remove gremlin.png"
        class="StyledButton-sc-323bzc-0 kQcSJk"
        type="button"
      >
        <svg
          aria-label="FormClose"
          class="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7 L17,17 M7,17 L17,7"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Dropzone renders file size and calculates correct suffix 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 icmjSc"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="StyledBox-sc-13pk1d4-0 eEjmTS"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eQTSO"
      >
        <span
          class="StyledText-sc-1sadyjn-0 bnqisW"
        >
          gremlin.png
        </span>
        <div
          class="StyledBox__StyledBoxGap-sc-13pk1d4-1 gufrXq"
        />
        <span
          class="StyledText-sc-1sadyjn-0 fWSbXS"
        >
          10.0
           
          KB
        </span>
      </div>
      <button
        aria-label="remove gremlin.png"
        class="StyledButton-sc-323bzc-0 kQcSJk"
        type="button"
      >
        <svg
          aria-label="FormClose"
          class="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7 L17,17 M7,17 L17,7"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Dropzone renders remove button 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 icmjSc"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="StyledBox-sc-13pk1d4-0 eEjmTS"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eQTSO"
      >
        <span
          class="StyledText-sc-1sadyjn-0 bnqisW"
        >
          gremlin.png
        </span>
      </div>
      <button
        aria-label="remove gremlin.png"
        class="StyledButton-sc-323bzc-0 kQcSJk"
        type="button"
      >
        <svg
          aria-label="FormClose"
          class="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7 L17,17 M7,17 L17,7"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Dropzone truncates text for long file name 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 icmjSc"
    tabindex="0"
  >
    <input
      accept=""
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="StyledBox-sc-13pk1d4-0 eEjmTS"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 eQTSO"
      >
        <span
          class="StyledText-sc-1sadyjn-0 bnqisW"
        >
          very-long-image...
        </span>
      </div>
      <button
        aria-label="remove very-long-image-title-that-should-be-truncated.png"
        class="StyledButton-sc-323bzc-0 kQcSJk"
        type="button"
      >
        <svg
          aria-label="FormClose"
          class="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7 L17,17 M7,17 L17,7"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
